modules/solid_mechanics/src/actions/DomainIntegralAction.C

```diff
@@ -51,14 +55,38 @@ DomainIntegralAction::DomainIntegralAction(const std::string & name, InputParame
   _have_crack_direction_vector_end_2(false),
   _treat_as_2d(getParam<bool>("2d")),
   _axis_2d(getParam<unsigned int>("axis_2d")),
-  _radius_inner(getParam<std::vector<Real> >("radius_inner")),
-  _radius_outer(getParam<std::vector<Real> >("radius_outer")),
   _convert_J_to_K(false),
   _has_symmetry_plane(isParamValid("symmetry_plane")),
   _symmetry_plane(_has_symmetry_plane ? getParam<unsigned int>("symmetry_plane") : std::numeric_limits<unsigned int>::max()),
   _position_type(getParam<MooseEnum>("position_type")),
+  _q_function_type(getParam<MooseEnum>("q_function_type")),
   _use_displaced_mesh(false)
 {
+  if (_q_function_type == GEOMETRY)
+  {
+    if (isParamValid("radius_inner") && isParamValid("radius_outer"))
+    {
+      _radius_inner = getParam<std::vector<Real> >("radius_inner");
+      _radius_outer = getParam<std::vector<Real> >("radius_outer");
+    }
+    else
+      mooseError("DomainIntegral error: must set radius_inner and radius_outer.");
+    for (unsigned int i=0; i<_radius_inner.size(); ++i)
+      _ring_vec.push_back(i+1);
+  }
+  else if (_q_function_type == TOPOLOGY)
+  {
+    if (isParamValid("ring_first") && isParamValid("ring_last"))
+    {
+      _ring_first = getParam<unsigned int>("ring_first");
+      _ring_last = getParam<unsigned int>("ring_last");
+    }
+    else
+      mooseError("DomainIntegral error: must set ring_first and ring_last if q_function_type = Topology.");
+    for (unsigned int i=_ring_first; i<=_ring_last; ++i)
+      _ring_vec.push_back(i);
+  }
+
```

Comment:
Maybe add a catch-all `else` case here, just in case you add a new `Q_FUNCTION_TYPE` enumeration somewhere down the line.